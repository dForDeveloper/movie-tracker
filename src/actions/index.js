export const addMovie = (movie) => ({
  type: 'ADD_MOVIE',
  movie
});

export const setUser = (user) => ({
  type: 'SET_USER',
  user
});

export const logoutUser = () => ({
  type: 'LOGOUT_USER'
});

export const getFavorites = (favorites) => ({
  type: 'GET_FAVORITES',
  favorites
});

export const addFavorite = (movie) => ({
  type: 'ADD_FAVORITE',
  movie
});