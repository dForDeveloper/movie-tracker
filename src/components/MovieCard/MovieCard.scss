.MovieCard--image {
  width: 80%;
}