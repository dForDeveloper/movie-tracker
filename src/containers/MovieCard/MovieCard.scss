.MovieCard {
  background-color: $opaque-white;
  border: 7px solid $border-color;
  border-radius: 10px;
  text-align: center;
  margin: 15px 0px;
  width: 80%;
  color: $body-font;
}

.moviecard-image {
  float: left;
  width: 100%;
}

.moviecard-title {
  display: inline-block;
  font-size: 26px;
  font-family: $heading-font;
  padding: 10px 10px;
}

.moviecard-favorite {
  float: left;
  padding: 5px;
  width: 100%;
  font-family: $body-font;
  font-size: 26px;
  font-weight: bold;
  background-color: rgba(243, 241, 255, 0.7);
  word-spacing: 20px;
  border: 3px solid $steel-grey;
}

.MovieCard:hover {
  box-shadow: 4px 4px 15px $off-white;
}

.MovieCard:hover h3 {
  color: white;
  text-shadow: $text-outline;
}

.moviecard-favorite:hover {
  background-image: linear-gradient(white 20%, $salmon 60%);
  cursor: pointer;
}