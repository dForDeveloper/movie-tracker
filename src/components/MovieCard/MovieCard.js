import React, { Component } from 'react';
import { connect } from 'react-redux';
import { fetchData } from '../../utils/api';
import { postFavorite, deleteFavorite } from '../../utils/queries';
import { getFavorites } from '../../actions';

class MovieCard extends Component {
  constructor() {
    super();
    this.state = {};
  }

  handleClick = async (user) => {
    if (user.name) {
      const movie = {
        ...this.props,
        user_id: user.id,
        movie_id: this.props.id
      };
      await this.toggleFavorite(movie);
    } else {
      // Prompt user to log in to save favorites
      return;
    }
  }

  toggleFavorite = async (movie) => {
    const { user_id, movie_id } = movie;
    if (this.props.favorite) {
      const response = await deleteFavorite(user_id, movie_id);
    } else {
      await postFavorite(movie);
    }
    const favorites = await fetchData(`
      http://localhost:3000/api/users/${user_id}/favorites
    `);
    this.props.getFavorites(favorites.data);
  }

  render() {
    const { title, poster_path, user, favorite } = this.props;
    return (
      <div className="MovieCard">
        <h3 className="movie-title">{title}</h3>
        <img
          src={`https://image.tmdb.org/t/p/w500${poster_path}`}
          alt={title}
          className='MovieCard--image'
        />
        <button onClick={() => this.handleClick(user)}>
          { favorite ? 'Remove from favorites' : 'Add to favorites'}
        </button>
      </div>
    );
  }
}

export const mapStateToProps = (state) => ({
  user: state.user
});

export const mapDispatchToProps = (dispatch) => ({
  getFavorites: (favorites) => dispatch(getFavorites(favorites))
});

export default connect(mapStateToProps, mapDispatchToProps)(MovieCard);