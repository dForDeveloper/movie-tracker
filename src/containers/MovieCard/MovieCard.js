import React, { Component } from 'react';
import { connect } from 'react-redux';
import { fetchData } from '../../utils/api';
import { setFavorites, toggleLoginPrompt } from '../../actions';

class MovieCard extends Component {
  constructor() {
    super();
    this.state = {};
  }

  handleClick = async (user) => {
    if (user.name) {
      const movie = {
        ...this.props,
        user_id: user.id,
        movie_id: this.props.id
      };
      await this.toggleFavorite(movie);
    } else {
      this.props.toggleLoginPrompt()
    }
  }

  toggleFavorite = async (movie) => {
    const { user_id, movie_id } = movie;
    if (this.props.favorite) {
      const url = `http://localhost:3000/api/users/${user_id}/favorites/${movie_id}`;
      const options = {
        method: 'DELETE',
        body: JSON.stringify({ user_id, movie_id }),
        headers: {
          'Content-Type': 'application/json'
        }
      }
      await fetchData(url, options);
    } else {
      const url = 'http://localhost:3000/api/users/favorites/new';
      const options = {
        method: 'POST',
        body: JSON.stringify(movie),
        headers: {
          'Content-Type': 'application/json'
        }
      };
      await fetchData(url, options);
    }
    const favesUrl = `http://localhost:3000/api/users/${user_id}/favorites`;
    const response = await fetchData(favesUrl);
    const favorites = response.data.map(favorite => favorite.movie_id);
    this.props.setFavorites(favorites);
  }

  render() {
    const { title, poster_path, currentUser, favorite } = this.props;
    return (
      <div className="MovieCard">
        <h3 className="movie-title">{title}</h3>
        <img
          src={`https://image.tmdb.org/t/p/w500${poster_path}`}
          alt={title}
          className='MovieCard--image'
        />
        <button onClick={() => this.handleClick(currentUser)}>
          { favorite ? 'Remove from favorites' : 'Add to favorites'}
        </button>
      </div>
    );
  }
}

export const mapStateToProps = (state) => ({
  currentUser: state.currentUser
});

export const mapDispatchToProps = (dispatch) => ({
  setFavorites: (favorites) => dispatch(setFavorites(favorites)),
  toggleLoginPrompt: () => dispatch(toggleLoginPrompt())
});

export default connect(mapStateToProps, mapDispatchToProps)(MovieCard);