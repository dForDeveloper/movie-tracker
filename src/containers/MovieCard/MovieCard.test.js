import React from 'react';
import { shallow } from 'enzyme';
import { MovieCard, mapStateToProps, mapDispatchToProps }  from './MovieCard';
import { setFavorites, toggleLoginPrompt } from '../../actions';

const mockProps = { 
  title: 'frozen',
  poster_path: 'url/imgur',
  currentUser: { id: 1, name: 'shannon' },
  favorite: true,
  id: 123456
}

describe('MovieCard', () => {
  let wrapper;
  beforeEach(() => {
    wrapper = shallow(
      <MovieCard {...mockProps}/>
    );
  });

  describe('MovieCard container', () => {
    it('should properly render the component elements', () => {
      expect(wrapper).toMatchSnapshot();
    });
  
    it('should call handleClick when a movie image is clicked', () => {
      wrapper.instance().handleClick = jest.fn();
      wrapper.update();
      wrapper.find('.moviecard-favorite').simulate('click');
      expect(wrapper.instance().handleClick).toBeCalled();
    });
  });

  describe('mapStateToProps', () => {
    it(`should return an object with a currentUser object`, () => {
      const mockState = {
        movies: [{ id: 234567, title: 'A Star is Born' }],
        currentUser: { id: 1, name: 'Jeo' },
        favorites: [],
        showLoginPrompt: false
      }
      const expected = { currentUser: { id: 1, name: 'Jeo' } };
      const result = mapStateToProps(mockState);
      expect(result).toEqual(expected);
    });
  });

  describe('mapDispatchToProps', () => {
    let dispatchMock = jest.fn()
    const result = mapDispatchToProps(dispatchMock);

    it('should call dispatch when toggleLoginPrompt is called', () => {
      const expected = toggleLoginPrompt(false);
      result.toggleLoginPrompt(false);
      expect(dispatchMock).toHaveBeenCalledWith(expected);
    });
  });
});