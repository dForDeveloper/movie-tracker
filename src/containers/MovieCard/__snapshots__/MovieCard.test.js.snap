// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MovieCard MovieCard container should properly render the component elements 1`] = `
<div
  className="MovieCard"
>
  <h3
    className="moviecard-title"
  >
    frozen
  </h3>
  <Link
    replace={false}
    to="/movies/123456"
  >
    <img
      alt="frozen"
      className="moviecard-image"
      src="https://image.tmdb.org/t/p/w500url/imgur"
    />
  </Link>
  <div
    className="moviecard-favorite"
    onClick={[Function]}
  >
    Remove ‚ùå
  </div>
</div>
`;
